Storage server contact & initial data refactor

jagerman opened the pull request:
This PR refactors how storage server tracks the contact information (ip, port, pubkeys) of network nodes and how it initiates contacts with other nodes on the network.

In particular, this aims to make new swarm joining more reliable: currently there are known issues where new swarm members are not receiving initial swarm messages because the other swarm members can send the data to the new node before the new node is ready to recognize them as members.

This reworks the mechanism:

When a new swarm member is detected, existing swarms members now send periodic (new) "snode.data_ready" requests to the new member(s), and do not attempt to push swarm data to that member until they get back a successful reply (thus indicating that we can reach the node, and that the node properly recognizes us as a swarm peer).  Previously members would attempt to push pretty much right away, but there are various ways the new node might either be unreachable, or might reject messages from a node it doesn't recognize.

If the command times out, or the remote returns anything other than "OK", the new member is returned to the queue and contact will be reattempted after another ~30s.

Once the OK is received, existing swarm members then push swarm messages to the new member via the existing snode.data endpoint for the new member to load the data.

All of this only applies if the remote has broadcast (via uptime proofs) that it is running a new enough storage server to support the handshake endpoint.

(This is not a perfect solution, and there are still various issues remaining, but this is a stop-gap while we also develop a more robust, active syncing mechanism).

Along the way this ends up touching a lot of other code:

- the "Swarms" class had a very confused triple purpose of tracking the current swarm, tracking all snodes, and tracking contact info, and was confusing to use.  These three purposes are now in three separate (though related) classes:
  - Swarm tracks the current swarm info, i.e. current swarm members.
  - Network tracks the global service node state, i.e. can look up which swarm a user pubkey or service node pubkey belongs to.
  - Contacts tracks the contact info of all registered service nodes, regardless of their state (aside from being registered).
  - sn_record is gone, and it *somewhat* replaced with a contact.  This isn't a direct replacement, though: many places that previously had a sn_record now simply take a legacy_pubkey, and look up the contact when needed via the Contacts object rather than passing the contact around.

- Storage server now subscribes to address updates from oxend (using https://github.com/oxen-io/oxen-core/pull/1820), so that it can update the contact info for a node immediately when a proof updates it, rather than only doing it when new blocks come in (which can be 2 minutes old during normal times, and much longer if there is a chain delay).

- We now fetch all SNs from oxend (previously only fully-funded ones): this lets us set up contacts and permissions for partially funded nodes, even though they aren't yet in any swarm, but then we are more likely to have them ready to go as soon as they get fully funded.

- Bump oxen-mq required version to 1.2.19, for an update that rechecks SN state on each command, rather than only when establishing a new connection.

- Replace all remaining std::stringstream usage with fmt.

- Changed the legacy_keypair (and similar for ed, x) from a std::pair typedef to a standalone struct with `.pub` and `.sec` members.  The plain pair was a bit dangerous because it wasn't immediately obvious whether .first was the pubkey or the privkey.  This also pushes it into some places that were still taking them as two separate arguments.
  - Related change: we now initialize from oxend with keypairs instead of just secret keys as we just had to compute all the pubkeys elsewhere anyway, so easier to just compute the pubkeys as soon as we get the secret keys.

- Added `load_from_bytes`/`load_from_hex` member functions to key types, and added `[[nodiscard]]` to the static `from_bytes`/`from_hex`. Previously it was easy to accidentally use `instance.from_hex(...)` which just calls the static method and doesn't actually change `instance`.

- Removed some string helper functions that are trivial now that we have fmt, added `get_human_readable_bytes` from oxen-core.

- Made the "random tests per tick" into a variable so that it can be changed easily later; previously the code was only ever trying 1.

- Replaced bt_list serialization with a more efficient bt_list_producer, along with a hack that lets us extract the final version-byte-prefixed string without needing extra allocations/copies.

- Remove the exclusive transaction from the save_bulk call (which was locked multiple times over): there doesn't seem to be any reason why this needs to be exclusive, blocking out other endpoints while the data gets loaded (which might be length as we could have 9MB of data to insert).

- Add some code documentation to better describe why we do what we do during swarm dissolution and new network swarms.

- Fetch the remote's storage server version information from oxend so that we can detect when the remote has updated to enable the handshake process.

- Removes version gates for old features that are always supported now.

jagerman committed:
Update oxen-encoding to latest

- Replaces long-deprecated and now-deleted `var::` calls for variant
  compatibility with `std::`
- Add missing headers for oxenc::to_hex
- add header for newer fmt compatibility

jagerman committed:
Update to de-networked libquic

Applies various small changes needed for upcoming libquic release.

jagerman committed:
Update storage seeds

Only public.loki.foundation was still correct; the others had moved to
different servers and had (unintended) oxend key changes that got missed
here.

This updates them to use the same three seeds that are used by Session
clients:

seed1, used directly (this machine is AKA "elsa", which is also the SPNS
    and big storage file server)
public-eu.optf.ngo == seed2
imaginary.stream == seed3

While we could use the seedN domain name for each, using separate TLDs
ensures some resilience against domain changes or expiry.

jagerman committed:
Bump oxenmq (1.2.20), oxenc (1.4.0)

oxenmq 1.2.19 has an important change in it that recalculates current SN
status on each command; previously it only did it on new connections,
which could cause authentication failures if a connection is established
when not an SN but then held open.  (1.2.20 includes some minor build
fixes).

oxenc 1.4.0 unbreaks some things that 1.3.0 broke.

jagerman committed:
Storage server contact & initial data refactor

This PR refactors how storage server tracks the contact information (ip,
port, pubkeys) of network nodes and how it initiates contacts with other
nodes on the network.

In particular, this aims to make new swarm joining more reliable:
currently there are known issues where new swarm members are not
receiving initial swarm messages because the other swarm members can
send the data to the new node before the new node is ready to recognize
them as members.

This reworks the mechanism:

When a new swarm member is detected, existing swarms members now send
periodic (new) "snode.data_ready" requests to the new member(s), and do
not attempt to push swarm data to that member until they get back a
successful reply (thus indicating that we can reach the node, and that
the node properly recognizes us as a swarm peer).  Previously members
would attempt to push pretty much right away, but there are various ways
the new node might either be unreachable, or might reject messages from
a node it doesn't recognize.

If the command times out, or the remote returns anything other than
"OK", the new member is returned to the queue and contact will be
reattempted after another ~30s.

Once the OK is received, existing swarm members then push swarm messages
to the new member via the existing snode.data endpoint for the new
member to load the data.

(This is not a perfect solution, and there are still various issues
remaining, but this is a stop-gap while we also develop a more robust,
active syncing mechanism).

Along the way this ends up touching a lot of other code:

- the "Swarms" class had a very confused triple purpose of tracking the
  current swarm, tracking all snodes, and tracking contact info, and was
  confusing to use.  These three purposes are now in three separate
  (though related) classes:
  - Swarm tracks the current swarm info, i.e. current swarm members.
  - Network tracks the global service node state, i.e. can look up which
    swarm a user pubkey or service node pubkey belongs to.
  - Contacts tracks the contact info of all registered service nodes,
    regardless of their state (aside from being registered).
  - sn_record is gone, and it *somewhat* replaced with a contact.  This
    isn't a direct replacement, though: many places that previously had
    a sn_record now simply take a legacy_pubkey, and look up the contact
    when needed via the Contacts object rather than passing the contact
    around.

- Storage server now subscribes to address updates from oxend, so that
  it can update the contact info for a node immediately when a proof
  updates it, rather than only doing it when new blocks come in (which
  can be 2 minutes old during normal times, and much longer if there is
  a chain delay).

- We now fetch all SNs from oxend (previously only fully-funded ones):
  this lets us set up contacts and permissions for partially funded
  nodes, even though they aren't yet in any swarm, but then we are more
  likely to have them ready to go as soon as they get fully funded.

- Bump oxen-mq required version to 1.2.19, for an update that rechecks
  SN state on each command, rather than only when establishing a new
  connection.

- Replace all remaining std::stringstream usage with fmt.

- Changed the legacy_keypair (and similar for ed, x) from a std::pair
  typedef to a standalone struct with `.pub` and `.sec` members.  The
  plain pair was a bit dangerous because it wasn't immediately obvious
  whether .first was the pubkey or the privkey.  This also pushes it
  into some places that were still taking them as two separate
  arguments.
  - Related change: we now initialize from oxend with keypairs instead
    of just secret keys as we just had to compute all the pubkeys
    elsewhere anyway, so easier to just compute the pubkeys as soon as
    we get the secret keys.

- Added `load_from_bytes`/`load_from_hex` member functions to key types,
  and added `[[nodiscard]]` to the static `from_bytes`/`from_hex`.
  Previously it was easy to accidentally use `instance.from_hex(...)`
  which just calls the static method and doesn't actually change
  `instance`.

- Removed some string helper functions that are trivial now that we have
  fmt, added `get_human_readable_bytes` from oxen-core.

- Made the "random tests per tick" into a variable so that it can be
  changed easily later; previously the code was only ever trying 1.

- Replaced bt_list serialization with a more efficient bt_list_producer,
  along with a hack that lets us extract the final version-byte-prefixed
  string without needing extra allocations/copies.

- Remove the exclusive transaction from the save_bulk call (which was
  locked multiple times over): there doesn't seem to be any reason why
  this needs to be exclusive, blocking out other endpoints while the
  data gets loaded (which might be length as we could have 9MB of data
  to insert).

- Add some code documentation to better describe why we do what we do
  during swarm dissolution and new network swarms.

jagerman committed:
Remove old HF version gates

jagerman committed:
Add SS version to contact; gate pre-transfer handshake on version

This adds retrieval and tracking of the remote storage server version to
the `contact` class so that we can use it when determining how we
interact with a peer.

This commit then uses that to gate our support for a pre-transfer
handshake on the remote running the new storage server version.

Also:
- bumps version to 2.10.0
- switches how we request service node info from the local oxend to use
  the simpler array-of-fields rather than the oxen-10 and earlier
  dict-of-true-values.

jagerman committed:
Extract storage version from addr change notifications

jagerman commented:
This is live and working on testnet, including going through some successful initial delayed handshake and them post-contact data pushes.

Doy-lee commented on oxenss/crypto/keys.h:
```diff
@@ -36,25 +35,31 @@ struct alignas(size_t) key_base : std::array<unsigned char, KeyLength> {
+    }
     // Same as above, but returns nullopt if invalid instead of throwing
-    static std::optional<Derived> maybe_from_hex(std::string_view hex) {
+    [[nodiscard]] static std::optional<Derived> maybe_from_hex(std::string_view hex) {
```
This function can be dropped since we don't use it.

Doy-lee commented on oxenss/http/http_client.cpp:
```diff
@@ -21,7 +21,8 @@ struct curl_context {
             client{client},
             sockfd{fd},
-            evt{event_new(client.loop->loop().get(), sockfd, 0, Client::curl_perform_c, this)} {}
+            evt{event_new(client.loop->get_event_base(), sockfd, 0, Client::curl_perform_c, this)} {
```
I think this can be slightly tidied by removing this and inlining to the point that we construct this in `handle_socket_c`. It puts all the event instantiation and adding to the event loop into one spot so there's less jumping around when trying to understand the nature of how `evt` is being used.

```cpp
event_del(curl_ctx->evt);
if (curl_ctx->evt) {
    event_assign(
            curl_ctx->evt,
            client.loop->get_event_base(),
            curl_ctx->sockfd,
            events,
            Client::curl_perform_c,
            curl_ctx);
} else {
    curl_ctx->evt = event_new(
            loop->get_event_base(),
            curl_ctx->sockfd,
            events,
            Client::curl_perform_c,
            curl_ctx);
}
event_add(curl_ctx->evt, NULL);
```
Doing that also gets rid of the need to have `friend curl_context` in the interface

jagerman replied:
While I'm not opposed to refactoring this code in such a way, I'm sort of ambivalent about making the change as suggested for a couple reasons: 1) it seems out of scope for what this PR is doing; and 2) currently `curl_context` is providing RAII mechanics (`event_new` and the corresponding `event_free` are only via curl_context construction/destruction) which this would break, and so I think such a change would need to go further by completely eliminating the curl_context class and bringing everything into handle_socket_c.  That's not necessarily a bad thing; it's just a bigger change than I want to make for this release.

Doy-lee commented on oxenss/rpc/request_handler.cpp:
```diff
@@ -41,42 +41,50 @@ static auto logcat = log::Cat("rpc");
+            snodes_json.push_back(json{
+                    // Deprecated; use pubkey_legacy instead:
+                    {"address", "{}.snode"_format(oxenc::to_base32z(snpk.view()))},
+                    // Deprecated string port for backwards compat; prefer https_port:
```
`port_https`

Doy-lee commented on oxenss/server/https.cpp:
```diff
@@ -284,17 +285,12 @@ namespace {
-                    result << '.';
-                result << +static_cast<uint8_t>(c);
-            }
+            result = "{:d}"_format(fmt::join(
```
`result = res.getRemoteAddressAsText()`?

This seems to also handle the 16 byte case but doesn't truncate leading 0s, potentially can just walk the string and eat the prefixed character until it meets a non `0` and is not a `:`.

jagerman replied:
I've switched it to simply use libquic's ipv4/ipv6 classes which have to_string() methods that do the right thing here (via inet_ntop).

Doy-lee commented on oxenss/snode/contacts.cpp:
```diff
@@ -0,0 +1,160 @@
+            changed = true;
+        }
+        if (c.pubkey_ed25519 && curr.pubkey_ed25519 != c.pubkey_ed25519) {
+            ed_pk.erase(curr.pubkey_ed25519);
```
Is there a reason this is not symmetric with the `x22519` branch?

```cpp
            if (curr.pubkey_x25519) {
                omq_rem.insert(curr.pubkey_x25519.str());
                x_pk.erase(curr.pubkey_x25519);
            }
```

The OMQ add/remove set can't be copied over because that's for `update_active_sns` which uses x25519 keys but how about the 0-key check?

jagerman replied:
Updated to be symmetric with the X25519 case.  (I think this was harmless as it'll just fail to find a 0 key to erase, but I agree it's better to be symmetric).

Doy-lee commented on oxenss/snode/contacts.h:
```diff
@@ -0,0 +1,193 @@
+    // element of [it, end), but more efficient as it does all updates in single lock.
+    //
+    // Returns the number of update() calls that return true.
+    template <contact_pair_iterator It, std::sentinel_for<It> End>
```
This isn't used, we _could_ delete it

jagerman replied:
It seems a bit lacking from an interface to omit it, even if currently unused (and just being an unsubstantiated template), so I've left it for now.

Doy-lee commented on oxenss/snode/network.cpp:
```diff
@@ -0,0 +1,97 @@
+    assert(bytes.size() == 32);
+
+    uint64_t res = 0;
+    for (size_t i = 0; i < 32; i += 8)
```
`i < bytes.size()`

Doy-lee commented on oxenss/snode/serialization.cpp:
```diff
@@ -14,43 +14,61 @@ namespace oxenss::snode {
+    //     ll[...][...][...]ee
+    //
+    // which we can then change the first `l` to \x01, and drop the final e, without needing
+    // to move or reallocate the string.
```
Smart

I think the only thing to improve here, but it's out of scope is to stream the messages we'll serialise to prevent having to load everything upfront into memory and on the side pre-reserve the required buffer here which would allow us to actually stay under the `SERIALIZATION_BATCH_SIZE` by knowing the payload size before appending.

jagerman replied:
Yes, absolutely, but I'll leave that for follow-up active syncing improvements.

Doy-lee commented on oxenss/snode/service_node.cpp:
```diff
@@ -380,19 +376,25 @@ void ServiceNode::send_onion_to_sn(
-            logcat, "Relaying data to: {} (x25519 pubkey {})", sn.pubkey_legacy, sn.pubkey_x25519);
+void ServiceNode::relay_data_reliable(
+        const std::string& blob, const crypto::legacy_pubkey& sn, const contact& ct) const {
+    if (!ct) {
```
IMO this can be removed by just inlining the method to the 1 call site that uses it. Then it can't be misused and so we don't need error checking. At most then we can just add an assert in case someone decides to copy-paste it elsewhere and forget to check.

jagerman replied:
Yeah, that's a good improvement since without that check it boils down to a single statement (not counting the debug log line).  I didn't add an assert because it seemed silly being two lines after a `if (ct && *ct)`.

Doy-lee commented on oxenss/snode/service_node.cpp:
```diff
@@ -724,6 +693,52 @@ void ServiceNode::update_swarms(std::promise<bool>* on_finish) {
 }
 
+void ServiceNode::process_snodes_update(std::string_view data) {
+    std::lock_guard lock{sn_mutex_};
```
Can move lock to after the parsing of data

jagerman replied:
Fixed.

Doy-lee commented on oxenss/snode/swarm.cpp:
```diff
@@ -1,297 +1,247 @@
-                const auto& snodes = swarm_info.snodes;
-                return std::find(snodes.begin(), snodes.end(), our_address_) != snodes.end();
-            });
+SwarmEvents Swarm::derive_swarm_events(const swarms_t& swarms) {
```
This one looks like it can be const function, doesn't seem to have side effects

jagerman replied:
Fixed.

Doy-lee reviewed (changes requested):
I'm not at a point where I can critically analyse and determine if this would introduce new bugs or issues to the current state of affairs but it looks fine as fine can be. 

General gist is populating the pending members queue, on a timer, mark those that are contactable, then relay to all of them and remove them from the pending member queue. Which sounds fine to me.

Once out of that queue, there's reasonably reliable syncing in place for when new messages come in. And still to be solved is when nodes drop off but come back online later.

jagerman committed:
Misc cleanups/improvements

From review suggestions:

- fix wrong name in comment https_port -> port_https
- drop unused maybe_from_hex
- replace manual ipv6 address stringification with libquic's
- update libquic checkout with some fixes/improvements for ipv4/ipv6
  handling needed by the above.
- make ed pubkey erasure skip null keys to be symmetric with x pubkey
  erasure code.  (The erase was harmless as a null key wouldn't get
  added, but the code symmetry is worthwhile).
- inline single-call relay_data_reliable because it is tiny and only
  ever called from one place.
- defer lock until after finished processing a swarm update, which
  doesn't need the lock.
- make derive_swarm_events const
